iree/compiler/Dialect/HAL/Conversion/FlowToHAL/ConvertStreamOps.cpp

```diff
@@ -540,6 +548,42 @@ static LogicalResult recordTensorClone(Value device, Value commandBuffer,
   return success();
 }
 
+/// Convert a `flow.tensor.reshape` to a copy if it is not an identity
+/// operation. If not an identity (i.e. it cannot be proved that the source
+/// buffer can be reused for the target buffer), then a copy is needed for
+/// correctness.
+static LogicalResult recordTensorReshape(Value device, Value commandBuffer,
+                                         IREE::Flow::TensorReshapeOp &reshapeOp,
+                                         BufferSet &bufferSet,
+                                         ConversionPatternRewriter &rewriter) {
+  if (isIdentityOp(reshapeOp)) return success();
+  auto &sourceBuffer = bufferSet.rangeMap[reshapeOp.source()];
+  auto &resultBuffer = bufferSet.rangeMap[reshapeOp.result()];
+
+  auto source = IREE::HAL::TensorRewriteAdaptor::getChecked(
+      reshapeOp.getLoc(), reshapeOp.source(), sourceBuffer.buffer, rewriter);
+  auto result = IREE::HAL::TensorRewriteAdaptor::getChecked(
+      reshapeOp.getLoc(), reshapeOp.result(), resultBuffer.buffer, rewriter);
+  if (!source.hasValue() || !result.hasValue()) {
+    return reshapeOp.emitOpError()
+           << "cannot create adaptors for tensor reshape operands/results";
```

Comment:
I'm confused that when do we use `rewriter.notifyMatchFailure()` and when do we use `op.emitOpError`.

I usually use `rewriter.notifyMatchFailure` when the pattern should handle it but this is not expected input or valid input. And use `return failure()` to choose the pattern to apply, e.g., i1 cases and non-i1 cases in StandardToSPIRV conversion.